turbine/src/xdp.rs

```diff
@@ -0,0 +1,185 @@
+    }
+}
+
+pub(crate) struct XdpSender {
+    senders: Vec<Sender<(Vec<SocketAddr>, shred::Payload)>>,
+}
+
+impl XdpSender {
+    #[inline]
+    pub(crate) fn try_send(
+        &self,
+        sender_index: u32,
+        addr: Vec<SocketAddr>,
+        payload: shred::Payload,
+    ) -> Result<(), TrySendError<(Vec<SocketAddr>, shred::Payload)>> {
+        self.senders[sender_index as usize % self.senders.len()].try_send((addr.clone(), payload))
+    }
+}
+
+pub(crate) struct XdpRetransmitter {
+    threads: Vec<thread::JoinHandle<()>>,
+}
+
+impl XdpRetransmitter {
+    #[cfg(not(target_os = "linux"))]
+    pub(crate) fn new(
+        _config: XdpConfig,
+        _src_port: u16,
+    ) -> Result<(Self, XdpSender), Box<dyn Error>> {
+        Err("XDP is only supported on Linux".into())
+    }
+
+    #[cfg(target_os = "linux")]
+    pub(crate) fn new(
+        config: XdpConfig,
+        src_port: u16,
+    ) -> Result<(Self, XdpSender), Box<dyn Error>> {
+        use caps::{
+            CapSet,
+            Capability::{CAP_BPF, CAP_NET_ADMIN, CAP_NET_RAW},
+        };
+
+        // switch to higher caps while we setup XDP. We assume that an error in
+        // this function is irrecoverable so we don't try to drop on errors.
+        for cap in [CAP_NET_ADMIN, CAP_NET_RAW, CAP_BPF] {
+            caps::raise(None, CapSet::Effective, cap)
+                .map_err(|e| format!("failed to raise {cap:?} capability: {e}"))?;
+        }
+
+        let dev = Arc::new(if let Some(interface) = config.interface {
+            NetworkDevice::new(interface).unwrap()
+        } else {
+            NetworkDevice::new_from_default_route().unwrap()
+        });
+
+        let ebpf = if config.zero_copy {
+            Some(
+                load_xdp_program(dev.if_index())
+                    .map_err(|e| format!("failed to attach xdp program: {e}"))?,
+            )
+        } else {
+            None
+        };
+
+        for cap in [CAP_NET_ADMIN, CAP_NET_RAW, CAP_BPF] {
+            caps::drop(None, CapSet::Effective, cap).unwrap();
+        }
+
+        let (senders, receivers) = (0..config.cpus.len())
+            .map(|_| crossbeam_channel::bounded(config.rtx_channel_cap))
+            .unzip::<_, _, Vec<_>, Vec<_>>();
+
+        let mut threads = vec![];
+
+        let (drop_sender, drop_receiver) = crossbeam_channel::bounded(1_000_000);
+        threads.push(
+            Builder::new()
+                .name("solRetransmDrop".to_owned())
+                .spawn(move || {
+                    loop {
+                        // drop shreds in a dedicated thread so that we never lock/madvise() from
+                        // the xdp thread
+                        match drop_receiver.try_recv() {
+                            Ok(i) => {
+                                drop(i);
+                            }
+                            Err(TryRecvError::Empty) => {
+                                thread::sleep(Duration::from_millis(1));
```

Comment:
Can the OS actually do anything meaningful with the 1ms sleep? Should we not embrace the dark side and just busy-wait this thread instead.

Reply:
No we shouldn't, have you seen the last screenshot here? https://github.com/anza-xyz/agave/pull/5671#issuecomment-2781861229

I'm not vibe coding this, assume that I've done the work.

Reply:
Yes, I have seen all the screenshots, yet I do not see how they drive the choice of specifically 1ms sleep here. I am assuming you are not vibe coding this, so if you have picked a sleep time of 1ms here, there must have been a good reason for this particular number. For instance, you could consider utilizing a blocking recv call here instead of a 1 ms sleep here, which would sleep until new objects are actually available, yet you have chosen not to do so. I am not sure what is the benefit of not explaining your reasoning really is.

Reply:
Ok just tested this in a toy example, the actual overhead here is pretty small, we can safely leave this particular bikeshed unpainted for now.

Reply:
Thank you for explaining this in detail! I agree it makes 100% sense to use this logic for the TX threads. 

However, for the drop thread here, do we still care about the latency enough to wake it up every milliscond?

Reply:
The reason to not sleep in the drop thread is that doing so would require a syscall in the xdp thread to wake it up, and in the xdp thread we don't want to be doing syscalls.